public/apps/configuration/utils/array-state-utils.tsx

```diff
@@ -0,0 +1,123 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { get, set, curry, StringRepresentable } from 'lodash';
+import { Dispatch, SetStateAction } from 'react';
+
+function resolveValue<T>(
+  value: T | (() => T)
+) {
+  return (typeof value === 'function') ? (value as Function)() : value;
+}
+
+/**
+ * Update an element in an sub array
+ * @param setStateCallback setState function
+ * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern" 
+ * @param newValue value to be updated
+ *
+ * e.g. 
+ *  currentState = [{a: 1}, {a: 2}]
+ *  path = [1, 'a']
```

Comment:
Minor: the index 1 causes some confusion with the value 1 above. Probably use index 0 as an example or add a comment at the end of the line saying that 1 is the array index.

Reply:
Done improvement.